Sources/SymbolKit/SymbolGraph/Symbol/KindIdentifier.swift

```diff
@@ -14,99 +14,124 @@ extension SymbolGraph.Symbol {
-            case .enum: return "enum"
-            case .case: return "enum.case"
-            case .func: return "func"
-            case .operator: return "func.op"
-            case .`init`: return "init"
-            case .ivar: return "ivar"
-            case .macro: return "macro"
-            case .method: return "method"
-            case .property: return "property"
-            case .protocol: return "protocol"
-            case .snippet: return "snippet"
-            case .snippetGroup: return "snippetGroup"
-            case .struct: return "struct"
-            case .subscript: return "subscript"
-            case .typeMethod: return "type.method"
-            case .typeProperty: return "type.property"
-            case .typeSubscript: return "type.subscript"
-            case .typealias: return "typealias"
-            case .var: return "var"
-            case .module: return "module"
-            case .unknown: return "unknown"
+            rawValue
+        }
+        
+        public static var allCases: Dictionary<String, Self>.Values {
+            _allCases.values
+        }
+        
+        private static var _allCases: [String: Self] = [
+            Self.associatedtype.rawValue: .associatedtype,
+            Self.class.rawValue: .class,
+            Self.deinit.rawValue: .deinit,
+            Self.enum.rawValue: .enum,
+            Self.case.rawValue: .case,
+            Self.func.rawValue: .func,
+            Self.operator.rawValue: .operator,
+            Self.`init`.rawValue: .`init`,
+            Self.ivar.rawValue: .ivar,
+            Self.macro.rawValue: .macro,
+            Self.method.rawValue: .method,
+            Self.property.rawValue: .property,
+            Self.protocol.rawValue: .protocol,
+            Self.snippet.rawValue: .snippet,
+            Self.snippetGroup.rawValue: .snippetGroup,
+            Self.struct.rawValue: .struct,
+            Self.subscript.rawValue: .subscript,
+            Self.typeMethod.rawValue: .typeMethod,
+            Self.typeProperty.rawValue: .typeProperty,
+            Self.typeSubscript.rawValue: .typeSubscript,
+            Self.typealias.rawValue: .typealias,
+            Self.var.rawValue: .var,
+            Self.module.rawValue: .module,
+            Self.extension.rawValue: .extension,
+        ]
+        
+        /// Register custom ``SymbolGraph/Symbol/KindIdentifier``s so they can be parsed correctly and
+        /// appear in ``allCases``.
+        ///
+        /// If a type is not registered, language prefixes cannot be removed correctly.
+        ///
+        /// - Note: When working in an uncontrolled environment where other parts of your executable might be disturbed by your
+        /// modifications to the symbol graph structure, register identifiers on your coder instead using
+        /// ``CustomizableCoder/register(symbolKinds:)`` and maintain your own list of ``allCases``.
+        public static func register(_ identifiers: Self...) {
```

Comment:
> Registering these at the static level on this model seems odd to me from an architecture perspective. It limits how SymbolKit can be used within the same client, e.g., if you're decoding symbol graphs from two different languages that use different kinds within the same client.

That's exactly the reason why I provide the alternative of registering the custom kinds to the Encoder/Decoder. However, I think that working with the static context is easier, **if** working in a controlled environment where there is only **one** "view" of the symbol graph format. `KindIdentifier` provides static functionality such as `allCases`, and I think this functionality should still work when extending the graph format.

>  Is there a reason we can't just parse the value as a string and just wrap it in a KindIdentifier value, without requiring clients to configure something ahead?

There might be some consequences for memory allocation as already discussed in #7, but I think the biggest problem is the `Equatable` conformance. For known kinds, SymbolKit removes the language specifier prefix, i.e. `swift.enum` is recognized as just `enum` and stored this way. Therefore, `KindIdentifier(rawValue: "enum") == KindIdentifier(identifier: "swift.enum")` yields **true** (note that the right initializer is used for decoding, whereas the left one is used for declaring the valid kinds in code). This is the behavior we want, as the language tag is handled elsewhere.

 If we don't know all valid kinds (because we don't allow custom kinds to be registered), we can't just throw away the prefix, as it might be important (as for `enum.case`). Thus, `swift.custom` must be stored as `swift.custom` and `KindIdentifier(rawValue: "custom") == KindIdentifier(identifier: "swift.custom")` yields **false**. This is very bad, if one wants to handle the kinds in a language agnostic way (as it is currently done in Swift-DocC).